clang/include/clang/Basic/TargetInfo.h

```diff
@@ -133,6 +133,7 @@ struct TransferrableTargetInfo {
   unsigned short SuitableAlign;
   unsigned short NewAlign;
   unsigned MaxVectorAlign;
```

Comment:
What's this used for? Can we limit it to 2^14?

Reply:
It's used in the same ASTContext code that's modified by this patch.  Probably fine to adjust it for x86.

Reply:
There's no special limitaion on x86. We can change ArgFlagsTy if the size is not the concern.

Reply:
`MaxVectorAlign` is set for x86 targets. I can use it as a limit as well, but note that this patch also makes indirect return value if size of the argument exceeds the limit. So, some changes can be caused to the current behavior and test changes like in for exampe `CodeGen/X86/x86-vec-i128.c`:

```
define{{.*}} <4 x i64> @test_v32u128(ptr noundef byval(<4 x i64>) align 32 %{{.*}}, ptr noundef byval(<2 x i128>) align 32 %{{.*}})
note: possible intended match here
define dso_local void @test_v32u128(ptr dead_on_unwind noalias writable sret(<4 x i64>) align 32 %agg.result, ptr noundef byval(<4 x i64>) align 32 %0, ptr noundef byval(<2 x i128>) align 32 %1)
``

If that is correct, I can use `MaxVectorAlign` for sure.

Reply:
I'm surprised the change of alignment affects ABI. Did you just expand
```
    MaxVectorAlign =
        hasFeature("avx512f") ? 512 : hasFeature("avx") ? 256 : 128;
```
to other targets in X86.h, or together with the change here?

I assumed the former won't change return type to `set`, though align may need to update.

Reply:
I did MaxVectorAlign = 2^14 in base TargetInfo, 
```   
 if (Ty->getAs<VectorType>() &&
        getContext().getTypeSize(Ty) > getTarget().getVectorAlignAlign())
      return getNaturalAlignIndirect(Ty);
```
in `ABIArgInfo X86_64ABIInfo::getIndirectReturnResult` like it is checked now with `getMaxPossibleAlign`.

Reply:
I think set `MaxVectorAlign = 2^14` in base TargetInfo only should solve the problem and won't result in test case failure.

Reply:
I'm not sure it will? With only modifying base TargetInfo clang will continue emit LLVM IR where giant vectors are returned directly, this code fails the verification.

Reply:
Right, sorry for the misleading. We still need to turn it to sret when vector size > 2^14. Then why you observed the changes in `CodeGen/X86/x86-vec-i128.c`? `<4 x i64>` is far less than the limitation.

Reply:
Because `MaxVectorAlign` is set to a smaller value for some x86 targets - https://github.com/llvm/llvm-project/blob/d8f0611acc2658ccc54d985044aa115716c6ad34/clang/lib/Basic/Targets/X86.h#L954 .

Reply:
I noticed that, but the value is set for Darwin, but the failure prefix is a non Darwin one.

Reply:
I pasted log in a weird way, sorry, all failures in this particular test I see are for Darwin.
```
// MEM256ALIGN16: define{{.*}} <4 x i64> @test_v32u128(ptr noundef byval(<4 x i64>) align 16 %{{.*}}, ptr noundef byval(<2 x i128>) align 16 %{{.*}})
                  ^
// MEM512ALIGN32: define{{.*}} <8 x i64> @test_v64u128(ptr noundef byval(<8 x i64>) align 32 %{{.*}}, ptr noundef byval(<4 x i128>) align 32 %{{.*}})
                  ^
```

Reply:
That makes sense. Passing/returning vector larger than target capacity (e.g. <4 x i64> without AVX) is ABI undefined behavior. The sret change matches with GCC, but may result in compatibility for Clang. Maybe we won't care it much.

Reply:
Ok, I updated the patch to use `MaxVectorAlign` instead. Thanks